### PR TITLE
auto reload config

### PR DESCRIPTION
**Pull request by F-TD5X**
#71 
I can't wait LOL, so I tried to impletment it myself.
I doesn't know more about Golang, so there might be some dirty code in this PR.

- in `core/inbound/server.go`, manually create servers for make them closeable.
- renamed `core/init.go` to `core/control.go`, and add some control code, **config watcher** code also wrote in this file.
- Let `config.Config` save it's path, for easy reload.

### CODE REVIEW

**Comment by F-TD5X** [comment]
Oh I forgot to rewatch changed file path. Hold on there.

**Comment by shawn1m** [comment]
Maybe it would be better only support something like "overture reloadconfig" and let the watcher outside ?

**Comment by F-TD5X** [comment]
You mean using some way to receive a `overture reloadconfig` signal? I don't know if golang has some cross platform way, or we have to open a port to do this communite.

**Comment by shawn1m** [comment]
Yeah, just like overture offers some apis for configuration change requests.

**Comment by F-TD5X** [comment]
`npipe` and `winio` cause unknown bugs on my windows vm, I gave up IPC, and use http to serve `/reload` request.

File: core/control.go [hunk 1]
```diff
@@ -36,7 +49,30 @@ func InitServer(configFilePath string) {
...
+}
+
+// Stop server
+func Stop() {
+	srv.Stop()
+}
+
+// ReloadHandler is passed to http.Server for handle "/reload" request
+func ReloadHandler(w http.ResponseWriter, r *http.Request) {
+	log.Warnf("Realod call received")
```
Comment (shawn1m) [inline core/control.go:65]: Warn level is not quite appropriate here

File: core/control.go [hunk 2]
```diff
@@ -36,7 +49,30 @@ func InitServer(configFilePath string) {
...
+	Reload()
+	io.WriteString(w, "Reloaded")
+}
 
-	s.Run()
+// Reload config and restart server
+func Reload() {
+	log.Warnf("Reloading")
```
Comment (shawn1m) [inline core/control.go:72]: Same as above

File: core/inbound/server.go [hunk 3]
```diff
@@ -104,23 +111,56 @@ func (s *Server) Run() {
 
 	for _, p := range [2]string{"tcp", "udp"} {
 		go func(p string) {
-			err := dns.ListenAndServe(s.bindAddress, p, mux)
+
+			// Manual create server inorder to have a way to close it.
+			srv := &dns.Server{Addr: s.bindAddress, Net: p, Handler: mux}
+			go func() {
+				<-s.ctx.Done()
+				log.Warnf("Shutting down the server on protocol %s", p)
```
Comment (shawn1m) [inline core/inbound/server.go:119]: Same as above

File: core/inbound/server.go [hunk 4]
```diff
@@ -104,23 +111,56 @@ func (s *Server) Run() {
...
+				srv.ShutdownContext(s.ctx)
+			}()
+			err := srv.ListenAndServe()
+			//err := dns.ListenAndServe(s.bindAddress, p, mux)
```
Comment (shawn1m) [inline core/inbound/server.go:123]: Remove it If it is not necessary

File: go.mod [hunk 5]
```diff
@@ -3,6 +3,7 @@ module github.com/shawn1m/overture
 go 1.12
 
 require (
+	github.com/fsnotify/fsnotify v1.4.9
```
Comment (shawn1m) [inline go.mod:6]: Remove them If it is not used in both go.mod and go.sum files

**Comment by F-TD5X** [comment]
sorry for delay for I have not received notificate email...

**Comment by shawn1m** [comment]
Merged. Thanks!

**Pull request merged**